File: platforms/windows/src/PalmierPro.Core/Models/BlendMode.cs
using System.Text.Json.Serialization;
using PalmierPro.Core.Json;

namespace PalmierPro.Core.Models;

/// How a visual clip composites over the layers below it. Normal = source-over.
[JsonConverter(typeof(SwiftStringEnumConverter<BlendMode>))]
public enum BlendMode
{
    [SwiftRawValue("normal")] Normal,
    [SwiftRawValue("darken")] Darken,
    [SwiftRawValue("multiply")] Multiply,
    [SwiftRawValue("colorBurn")] ColorBurn,
    [SwiftRawValue("lighten")] Lighten,
    [SwiftRawValue("screen")] Screen,
    [SwiftRawValue("colorDodge")] ColorDodge,
    [SwiftRawValue("overlay")] Overlay,
    [SwiftRawValue("softLight")] SoftLight,
    [SwiftRawValue("hardLight")] HardLight,
    [SwiftRawValue("difference")] Difference,
    [SwiftRawValue("exclusion")] Exclusion,
    [SwiftRawValue("hue")] Hue,
    [SwiftRawValue("saturation")] Saturation,
    [SwiftRawValue("color")] Color,
    [SwiftRawValue("luminosity")] Luminosity,
}

public static class BlendModeExtensions
{
    public static string DisplayName(this BlendMode mode) => mode switch
    {
        BlendMode.Normal => "Normal",
        BlendMode.Darken => "Darken",
        BlendMode.Multiply => "Multiply",
        BlendMode.ColorBurn => "Color Burn",
        BlendMode.Lighten => "Lighten",
        BlendMode.Screen => "Screen",
        BlendMode.ColorDodge => "Color Dodge",
        BlendMode.Overlay => "Overlay",
        BlendMode.SoftLight => "Soft Light",
        BlendMode.HardLight => "Hard Light",
        BlendMode.Difference => "Difference",
        BlendMode.Exclusion => "Exclusion",
        BlendMode.Hue => "Hue",
        BlendMode.Saturation => "Saturation",
        BlendMode.Color => "Color",
        BlendMode.Luminosity => "Luminosity",
        _ => throw new ArgumentOutOfRangeException(nameof(mode)),
    };
}

File: Sources/PalmierPro/Models/BlendMode.swift
import Foundation

/// How a visual clip composites over the layers below it. `normal` = source-over.
enum BlendMode: String, Codable, Sendable, CaseIterable {
    case normal, darken, multiply, colorBurn, lighten, screen, colorDodge
    case overlay, softLight, hardLight, difference, exclusion
    case hue, saturation, color, luminosity

    var displayName: String {
        switch self {
        case .normal: L10n.string("Normal")
        case .darken: L10n.string("Darken")
        case .multiply: L10n.string("Multiply")
        case .colorBurn: L10n.string("Color Burn")
        case .lighten: L10n.string("Lighten")
        case .screen: L10n.string("Screen")
        case .colorDodge: L10n.string("Color Dodge")
        case .overlay: L10n.string("Overlay")
        case .softLight: L10n.string("Soft Light")
        case .hardLight: L10n.string("Hard Light")
        case .difference: L10n.string("Difference")
        case .exclusion: L10n.string("Exclusion")
        case .hue: L10n.string("Hue")
        case .saturation: L10n.string("Saturation")
        case .color: L10n.string("Color")
        case .luminosity: L10n.string("Luminosity")
        }
    }

    /// Core Image blend-filter name; nil for `normal` (plain source-over compositing).
    var ciFilterName: String? {
        switch self {
        case .normal: nil
        case .darken: "CIDarkenBlendMode"
        case .multiply: "CIMultiplyBlendMode"
        case .colorBurn: "CIColorBurnBlendMode"
        case .lighten: "CILightenBlendMode"
        case .screen: "CIScreenBlendMode"
        case .colorDodge: "CIColorDodgeBlendMode"
        case .overlay: "CIOverlayBlendMode"
        case .softLight: "CISoftLightBlendMode"
        case .hardLight: "CIHardLightBlendMode"
        case .difference: "CIDifferenceBlendMode"
        case .exclusion: "CIExclusionBlendMode"
        case .hue: "CIHueBlendMode"
        case .saturation: "CISaturationBlendMode"
        case .color: "CIColorBlendMode"
        case .luminosity: "CILuminosityBlendMode"
        }
    }
}
